src/Orleans.Pipeline.Client/OrleansPipe.cs

```diff
@@ -64,48 +65,68 @@ public async Task Stop(CancellationToken token)
         }
         catch (Exception ex)
         {
-            logger.LogError(ex, "An error occurred during Stopping");
+            logger.LogError(ex, "An error occurred during stopping.");
         }
         finally
         {
-            _writerStoppingTokenSource.Dispose();
             _observerStoppingTokenSource.Dispose();
         }
     }
 
     #endregion
 
     #region IOrleansPipeObserver<TFromServer>
+
     public async Task ReceiveMessage(PipeTransferItem<TFromServer> message)
     {
-        if (message is { Mode: TransferMode.Data, Item: not null })
+        switch (message.Mode)
         {
-            await _reader.Writer.WriteAsync(message.Item);
+            case TransferMode.Data:
+                {
+                    ArgumentNullException.ThrowIfNull(message.Item, "Item must no be null using data transfer mode");
+                    await _reader.Writer.WriteAsync(message.Item);
+                }
+                break;
+            case TransferMode.Heartbeat: 
+                _lastHeartbeat = DateTime.UtcNow;
+                break;
+            default:
+                throw new NotSupportedException($"The transfer mode '{message.Mode}' is not supported");
         }
     }
+
     #endregion
 
     private async Task RunWriter(CancellationToken cancellationToken)
     {
         try
         {
-            using var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _writerStoppingTokenSource.Token);
-            await foreach (var result in _writer.Reader.ReadAllAsync(linkedCts.Token))
+            await foreach (var result in _writer.Reader.ReadAllAsync(cancellationToken))
             {
+                if ((DateTime.UtcNow - _lastHeartbeat) > ExpectedHeartbeatInterval)
+                {
+                    //TODO: Continue
+                }
+
```

Comment:
If we do this here we only know things are/were broken when we receive the next item. this would be a bit late (if i did understand the changes correctly).

Since its with a TODO here, its more a question.

Reply:
no you're right, i was just in a rush